monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py

```diff
@@ -0,0 +1,75 @@
+from multiprocessing import Queue, get_context
+from unittest.mock import MagicMock
+
+import pytest
+from tests.unit_tests.infection_monkey.base_island_api_client import BaseIslandAPIClient
+
+from common import OperatingSystem
+from infection_monkey.exploit import CachingAgentBinaryRepository
+from infection_monkey.island_api_client import IIslandAPIClient
+
+AGENT_BINARY = b"BINARY_EXECUTABLE"
+
+
+@pytest.fixture
+def mock_island_api_client() -> IIslandAPIClient:
+    return StubIslandAPIClient()
+
+
+class PickableMagicMock(MagicMock):
+    def __reduce__(self):
+        return (MagicMock, ())
+
+
+class StubIslandAPIClient(BaseIslandAPIClient):
+    def __init__(self):
+        self._calls = get_context("spawn").Value("i", 0)
+
+    @property
+    def calls(self):
+        return self._calls.value
+
+    def get_agent_binary(self, _: OperatingSystem) -> bytes:
+        with self._calls.get_lock():
+            self._calls.value += 1
+            print(f"CALLED: {self._calls}")
+            return AGENT_BINARY
+
+
+def test_get_agent_binary__uses_cache(mock_island_api_client):
```

Comment:
It's also worth adding a test that if I call get() for 2 different operating systems, I get different binaries back.

Reply:
No problemo!